tensorflow/java/src/main/java/org/tensorflow/op/NameScope.java

```diff
@@ -0,0 +1,148 @@
+ */
+final class NameScope {
+
+  NameScope withSubScope(String scopeName) {
+    checkPattern(NAME_REGEX, scopeName);
+
+    // Override with opName if it exists.
+    String actualName = (opName != null) ? opName : scopeName;
+
+    String newPrefix = fullyQualify(makeUnique(actualName));
+    return new NameScope(newPrefix, null, null);
+  }
+
+  NameScope withName(String name) {
+    checkPattern(NAME_REGEX, name);
+
+    // All context except for the opName is shared with the new scope.
+    return new NameScope(opPrefix, name, ids);
+  }
+
+  String makeOpName(String name) {
+    checkPattern(NAME_REGEX, name);
+
+    // Override with opName if it exists.
+    String actualName = (opName != null) ? opName : name;
+
+    return fullyQualify(makeUnique(actualName));
+  }
+
+  /**
+   * Create a new, root-level namescope.
+   *
+   * <p>A root-level namescope generates operator names with no components, like {@code Const_72}
+   * and {@code result}.
+   */
+  NameScope() {
+    this(null, null, null);
+  }
+
+  private NameScope(String opPrefix, String opName, Map<String, Integer> ids) {
+    this.opPrefix = opPrefix;
+    this.opName = opName;
+    if (ids != null) {
+      this.ids = ids;
+    } else {
+      this.ids = new HashMap<String, Integer>();
+    }
+  }
+
+  // Generate a unique name, different from existing ids.
+  //
+  // ids is a map from id to integer, representing a counter of the
+  // number of previous requests to generate a unique name for the
+  // given id.
+  //
+  // For instance, the first use of makeUnique("a") adds "a" -> 1
+  // to ids and returns "a".
+  //
+  // The second use of makeUnique("a") updates ids to "a" -> 2
+  // and returns "a_1", and so on.
+  private String makeUnique(String id) {
+    if (!ids.containsKey(id)) {
+      ids.put(id, 1);
+      return id;
+    } else {
+      int cur = ids.get(id);
+      ids.put(id, cur + 1);
+      return String.format("%s_%d", id, cur);
+    }
+  }
+
+  private String fullyQualify(String name) {
+    if (opPrefix != null) {
+      return String.format("%s/%s", opPrefix, name);
+    } else {
+      return name;
+    }
+  }
+
+  // If opPrefix is non-null, it is a prefix applied to all names
+  // created by this instance.
+  private final String opPrefix;
+
+  // If opName is non-null, it is used to derive the unique name
+  // for operators rather than the provided default name.
+  private final String opName;
+
+  // NameScope generates unique names by appending a numeric suffix if
+  // needed. This is a map containing names already created by this
+  // instance mapped to the next available numeric suffix for it.
+  private final Map<String, Integer> ids;
+
+  private static void checkPattern(Pattern pattern, String name) {
+    if (name == null) {
+      throw new IllegalArgumentException("Names cannot be null");
+    }
+    if (!pattern.matcher(name).matches()) {
+      throw new IllegalArgumentException(String.format("Invalid name '%s'", name));
```

Comment:
Let's add some more detail to this error, perhaps something like:

"invalid name: '%s' does not match the regular expression NAME_REGEX"